### PR TITLE
Nexus: Further updates to `PseudoSet`, adding `generate_pseudoset`

### PR DESCRIPTION
**Pull request by brockdyer03**
## Proposed changes

This PR introduces the user-facing function `generate_pseudoset` that behaves similarly to the existing `generate`-style functions in Nexus, and acts as a replacement for the now-deprecated `ppset`. It additionally takes the functionality of `PseudoSet.pseudo_remap()` and merges it into the existing `PseudoSet.get_pseudos()`.

## What type(s) of changes does this code introduce?

- New feature
- Testing changes (e.g. new unit/integration/performance tests)
- Documentation or build script changes

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

Laptop, Fedora Linux 43 (KDE Plasma Desktop Edition)
AMD Ryzen 7 PRO 7840U (8 cores, 16 logical processors)
```
Python        3.14.6
uv            0.12.5
cif2cell      2.1.0
coverage      7.15.4
h5py          3.16.0
matplotlib    3.11.1
numpy         2.5.2
pycifrw       4.4.6
pydot         4.0.1
pytest        9.1.1
pytest-cov    7.1.0
pytest-order  1.5.0
scipy         1.18.0
seekpath      2.2.1
spglib        2.7.0
sphinx        9.1.0
```

## Checklist

* * [ ] I have read the pull request guidance and develop docs
* * [ ] This PR is up to date with the current state of 'develop'
* * [ ] Code added or changed in the PR has been clang-formatted
* * [ ] This PR adds tests to cover any new code, or to catch a bug that is being fixed
* * [ ] Documentation has been added (if appropriate)

### CODE REVIEW

File: nexus/nexus/pseudoset.py [hunk 1]
```diff
@@ -234,30 +297,75 @@ def read_potcar_z_valence(file: PathLike) -> int | float:
...
+    label : str
+        Label of the pseudopotential set.
+    **codes_pps
+        Map of codes to collections of names of the pseudopotential files.
+
+    Notes
+    -----
+    Using this function is generally discouraged for users as the use of
+    a string label for identifying pseudopotentials prevents Python from
+    raising a :exc:`SyntaxError` when the label supplied to later
```
Comment (jtkrogel) [inline nexus/nexus/pseudoset.py]: This should simply be fixed.  It is not an intrinsic issue.

File: nexus/nexus/pseudoset.py [hunk 2]
```diff
@@ -309,107 +415,65 @@ class PseudoSet(DevBase):
         # https://www.quantum-espresso.org/Doc/INPUT_PW.html#id268
         "espresso": frozenset({".ncpp", ".upf", ".vdb", ".van", ".rrkj3"}),
```
Comment (jtkrogel) [inline nexus/nexus/pseudoset.py:416]: This needs to be the full package name here and elsewhere.

File: nexus/nexus/pseudoset.py [hunk 3]
```diff
@@ -309,107 +415,65 @@ class PseudoSet(DevBase):
...
                     if skip_invalid:
                         warn(msg)
                         continue
                     else:
                         raise FileNotFoundError(msg)
-                else:
-                    # No need to check if `label` is already defined since
-                    # dictionary keys are, by definition, unique.
-                    self.pseudos[label] = psp
+                #end if psp.exists()
```
Comment (jtkrogel) [inline nexus/nexus/pseudoset.py]: Just `#end if`

File: nexus/nexus/pseudoset.py [hunk 4]
```diff
@@ -309,107 +415,65 @@ class PseudoSet(DevBase):
...
+                elif psp.name in PseudoSet.pseudo_files:
+                    psp = Path(PseudoSet.pseudo_files[psp.name]).resolve()
+                else:
                     msg = f"Pseudo file {psp} can not be located"
                     if skip_invalid:
                         warn(msg)
                         continue
                     else:
                         raise FileNotFoundError(msg)
+                #end if psp.exists()
```
Comment (jtkrogel) [inline nexus/nexus/pseudoset.py]: Same

File: nexus/nexus/pseudoset.py
Comment (jtkrogel) [inline nexus/nexus/pseudoset.py]: This should not be a warning.  Any code name appearing in `generate_*` must be a valid key.

File: nexus/nexus/pseudoset.py [hunk 5]
```diff
@@ -532,14 +596,32 @@ def _check_code_str(code: str) -> Literal["espresso", "gamess", "vasp", "qmcpack
...
 
 
+    @staticmethod
+    def _normalize_code_map_keys(mapping: Mapping) -> dict:
+        """Take a dict with any code keys and normalize them.
+
+        Normalizing in this case means checking the keys and making sure they
+        match those in :attr:`PseudoSet.file_exts`.
+        """
+        mapping = deepcopy(dict(mapping))
```
Comment (jtkrogel) [inline nexus/nexus/pseudoset.py:614]: Conversion to dict unnecessary

File: nexus/nexus/pseudoset.py [hunk 6]
```diff
@@ -532,14 +596,32 @@ def _check_code_str(code: str) -> Literal["espresso", "gamess", "vasp", "qmcpack
...
+        for code in list(mapping):
```
Comment (jtkrogel) [inline nexus/nexus/pseudoset.py]: Conversion to list unnecessary.  Use `for code in mapping.keys()` for clarity.

File: nexus/nexus/pseudoset.py [hunk 7]
```diff
@@ -1075,4 +1355,191 @@ def __repr__(self) -> str:
         rep += ")\n"
         return rep
     #end def __repr__
-#end class PseudoSet
\ No newline at end of file
+#end class PseudoSet
+
+
+def generate_pseudoset(
+    pseudo_dir: PathLike | None = None,
```
Comment (jtkrogel) [inline nexus/nexus/pseudoset.py:1422]: Within this function at least, `pseudo_dir` needs to default to the one provided via `settings`

File: nexus/nexus/pseudoset.py [hunk 8]
```diff
@@ -1075,4 +1355,191 @@ def __repr__(self) -> str:
...
+    *,
+    code       : str | Collection[str] | None = None,
+    extension  : Mapping[str, str | Collection[str]] | None = None,
+    pattern    : Mapping[str, str | Pattern] | str | None = None,
```
Comment (jtkrogel) [inline nexus/nexus/pseudoset.py]: This name is too generic for UI (it represents a type, not a descriptive name).  As we discussed (and agreed) earlier, the name should be `include`.

Also, this should accept `glob` (shell style) as well as regex.  It is much more familiar to people, and they will try things like `'*ccECP*'`.  This should work for those common cases (shell cases in general).

File: nexus/nexus/pseudoset.py [hunk 9]
```diff
@@ -1075,4 +1355,191 @@ def __repr__(self) -> str:
         rep += ")\n"
         return rep
     #end def __repr__
-#end class PseudoSet
\ No newline at end of file
+#end class PseudoSet
+
+
+def generate_pseudoset(
```
Comment (jtkrogel) [inline nexus/nexus/pseudoset.py:1421]: Since this function is in the main UI, it needs documentation in the user manual.  The few main call signatures (not lots of fancy ones) should be shown.  The interface with the simulation generation functions should also be shown, alongside functioning original "plain list of pseudopotential filenames" case.

File: nexus/nexus/pseudoset.py [hunk 10]
```diff
@@ -904,79 +1016,248 @@ def from_mixed_dir(
...
     #end def from_mixed_dir
 
 
-    def get_pseudos(
+    def _get_pseudos(
         self,
-        system: PhysicalSystem | Iterable[str],
-        code  : Literal["espresso", "gamess", "vasp", "qmcpack", "rmg", "pyscf"] | None = None,
-        ) -> set[Path] | None:
+        system: PhysicalSystem | Collection[str],
```
Comment (jtkrogel) [inline nexus/nexus/pseudoset.py:1097]: Please include a functioning test that accepts `system=None` without raising errors.  There are simulation cases that do not (and should not) require a PhysicalSystem to be provided (e.g. pwscf with native system generation in its input file).

File: nexus/nexus/pseudoset.py (hunk 1 above)
Comment (jtkrogel) [reply nexus/nexus/pseudoset.py]: You don't have to fix it here.  If you prefer, I will make a fix in a follow-on PR.

**Review by jtkrogel — CHANGES_REQUESTED** [review]
Most significant changes relate to the UI level + manual docs.

File: nexus/nexus/pseudoset.py (hunk 1 above)
Comment (brockdyer03) [reply nexus/nexus/pseudoset.py]: This can't be fixed, it's the whole concept of `ppset`

File: nexus/nexus/pseudoset.py
Comment (brockdyer03) [reply nexus/nexus/pseudoset.py]: The point of the function is to normalize code names to a known set. I am fine not making it a warning, but I still will be changing the name to `quantum_espresso` (of course, after I update the existing `espresso` to `quantum_espresso`)

File: nexus/nexus/pseudoset.py (hunk 5 above)
Comment (brockdyer03) [reply nexus/nexus/pseudoset.py:614]: It is necessary since if I don't then an immutable datatype causes this to fail (e.g. `MappingProxyType` or, in Python 3.15, `frozendict`)

File: nexus/nexus/pseudoset.py (hunk 6 above)
Comment (brockdyer03) [reply nexus/nexus/pseudoset.py]: If I don't make it list then I get an error that the dict is modified during iteration.

File: nexus/nexus/pseudoset.py (hunk 7 above)
Comment (brockdyer03) [reply nexus/nexus/pseudoset.py:1422]: It does already do that by default, it just occurs in the backend.

File: nexus/nexus/pseudoset.py (hunk 8 above)
Comment (brockdyer03) [reply nexus/nexus/pseudoset.py]: I think that involves quite a lot more work on the backend than you may expect. Is it okay if that is a separate PR?

File: nexus/nexus/pseudoset.py (hunk 9 above)
Comment (brockdyer03) [reply nexus/nexus/pseudoset.py:1421]: Can I do that in a separate PR, the same one I change `pattern` to `include` and `exclude`?

File: nexus/nexus/pseudoset.py (hunk 10 above)
Comment (brockdyer03) [reply nexus/nexus/pseudoset.py:1097]: It already is fine with `None`, I just forgot to change this signature. It gets tested like that in `test_rmg_input.py::test_generate`. Also, to be fair, the `Collection[str]` would be the ideal thing to use in the case where someone passes `elem`, or in the case of the RMG format that `generate` function could pass the elements in `atoms`.

File: nexus/nexus/pseudoset.py (hunk 2 above)
Comment (brockdyer03) [reply nexus/nexus/pseudoset.py:416]: I want to push back on this. The problem I have encountered with making this the full package name is that it is simply a very long name. This means that it is often going to be more convenient to use any available aliases, which defeats the purpose of using the full name. Something like Quantum Package does require the full name though because of the ambiguity of any of the individual words, but there don't appear to be any other programs in the scope of Nexus with `espresso` in the name.

I additionally do not want to just use `pwscf` like previous code has, for several reasons. Primarily, it is too similar to `pyscf` and can lead to confusion ***at a glance***. More importantly though, Quantum ESPRESSO uses a unified pseudopotential format across all of its internal code, so at least for pseudopotentials it is misleading to use `pwscf` as the internal name.

File: nexus/nexus/pseudoset.py (hunk 8 above)
Comment (brockdyer03) [reply nexus/nexus/pseudoset.py]: I've decided to only do `glob` style patterns. It's not possible to tell if a string is a regex string or a `glob` pattern a-priori, and I don't think it's wise to add a toggle option between them since it just complicates the already extensive call signature. Since regex is already a dark art, it feels more reasonable to just ditch it entirely.

File: nexus/nexus/pseudoset.py (hunk 2 above)
Comment (jtkrogel) [reply nexus/nexus/pseudoset.py:416]: If it appears in `generate_*` for a simulation, that code (by name) must be one of the available options.  It is good to also supply `quantum_espresso`.  Abbreviations can also be fine in addition, e.g. `qe` -- but in addition.

File: nexus/nexus/pseudoset.py (hunk 8 above)
Comment (jtkrogel) [reply nexus/nexus/pseudoset.py]: Sounds good

File: nexus/nexus/pseudoset.py (hunk 9 above)
Comment (jtkrogel) [reply nexus/nexus/pseudoset.py:1421]: Sure

File: nexus/nexus/pseudoset.py (hunk 10 above)
Comment (jtkrogel) [reply nexus/nexus/pseudoset.py:1097]: OK, I will look

File: nexus/nexus/pseudoset.py (hunk 2 above)
Comment (brockdyer03) [reply nexus/nexus/pseudoset.py:416]: I'm fine supporting other things, but the internal normalized name is what I am mostly concerned with.

File: nexus/nexus/pseudoset.py (hunk 1 above)
Comment (jtkrogel) [reply nexus/nexus/pseudoset.py]: Noting that I will make a follow-on PR to fix.

File: nexus/nexus/pseudoset.py (hunk 2 above)
Comment (jtkrogel) [reply nexus/nexus/pseudoset.py:416]: Of course.  The normalized internal name should be the package name: `quantum_espresso`.  You can have other aliases by preference, e.g. `espresso` or `qe`.  Alternate selections (including `pwscf`) should not raise warnings.  The way to handle preferred behavior is to only provide manual examples in the preferred style (i.e. `quantum_espresso` or `qe`).

**Review by jtkrogel — CHANGES_REQUESTED** [review]
Small default/accepted key issue remaining.  Test discussion to occur separately.

File: nexus/nexus/pseudoset.py (hunk 2 above)
Comment (brockdyer03) [reply nexus/nexus/pseudoset.py:416]: I think a warning is appropriate since users may try to access `pwscf` later on and might not notice that the normalization happened at all. I won't deprecate the aliases though.

**Review by jtkrogel — APPROVED** [review]
Good now
